### PR TITLE
scheduler: new hotspot scheduler basic part

### PR DESCRIPTION
**Pull request by Luffbee**
### What problem does this PR solve? 

Currently, the hotspot scheduler use the number of hot regions to make schedule decision, which cannot cover "many warm region" case, and may make wrong schedule decisions.

### What is changed and how it works?

Use store flow byte rate to make decision.

### Check List 

Tests 

 - Unit test

Related changes

 - Need to be included in the release notes

### CODE REVIEW

File: server/statistics/store.go [hunk 1]
```diff
@@ -65,7 +65,24 @@ func (s *StoresStats) GetRollingStoreStats(storeID uint64) *RollingStoreStats {
...
+	if !ok {
+		store = newRollingStoreStats()
+		s.rollingStoresStats[storeID] = store
+	}
+	store.Observe(stats)
+}
+
+// Set sets the store statistics (for test).
+func (s *StoresStats) Set(storeID uint64, stats *pdpb.StoreStats) {
+	s.RLock()
```
Comment (nolouch) [inline server/statistics/store.go]: Lock()? `Observe` have the same  problem.  and Why not use Observe directly?

Comment (Luffbee) [reply server/statistics/store.go]: `Observe`  `Add`s the new data into the `MovingAvg`, which will influenced by history. This one `Set`s the `MovingAvg` to the new data.

**Comment by nolouch** [comment]
PTAL @rleungx

File: server/schedulers/hot_test.go [hunk 2]
```diff
@@ -52,84 +53,134 @@ func (s *testHotWriteRegionSchedulerSuite) TestSchedule(c *C) {
...
-	// Report store written bytes.
+	//| store_id | write_bytes_rate |
+	//|----------|------------------|
+	//|    1     |       7.5MB      |
+	//|    2     |       4.5MB      |
+	//|    3     |       4.5MB      |
+	//|    4     |        6MB       |
+	//|    5     |        0MB       |
+	//|    6     |        0MB       |
 	tc.UpdateStorageWrittenBytes(1, 75*1024*1024)
```
Comment (rleungx) [inline server/schedulers/hot_test.go]: I prefer to define some constants like `KB`, `MB`.

File: server/statistics/store.go [hunk 3]
```diff
@@ -63,22 +63,43 @@ func (s *StoresStats) GetRollingStoreStats(storeID uint64) *RollingStoreStats {
...
-	s.RLock()
-	defer s.RUnlock()
-	s.rollingStoresStats[storeID].Observe(stats)
+	s.Lock()
+	store, ok := s.rollingStoresStats[storeID]
+	if !ok {
+		store = newRollingStoreStats()
+		s.rollingStoresStats[storeID] = store
+	}
+	s.Unlock()
```
Comment (rleungx) [inline server/statistics/store.go]: How about moving these lines into a function?

File: server/schedulers/hot_region.go [hunk 4]
```diff
@@ -230,17 +231,24 @@ func (h *balanceHotRegionsScheduler) balanceHotWriteRegions(cluster opt.Cluster)
...
 
-func calcScore(storeItems map[uint64][]*statistics.HotPeerStat, cluster opt.Cluster, kind core.ResourceKind) statistics.StoreHotRegionsStat {
+func calcScore(storeHotPeers map[uint64][]*statistics.HotPeerStat, storeBytesStat map[uint64]float64, cluster opt.Cluster, kind core.ResourceKind) statistics.StoreHotRegionsStat {
 	stats := make(statistics.StoreHotRegionsStat)
-	for storeID, items := range storeItems {
-		// HotDegree is the update times on the hot cache. If the heartbeat report
-		// the flow of the region exceeds the threshold, the scheduler will update the region in
-		// the hot cache and the hotdegree of the region will increase.
+	for storeID, items := range storeHotPeers {
+		storeHots, ok := stats[storeID]
```
Comment (rleungx) [inline server/schedulers/hot_region.go]: `storeHots` sounds not good enough?

Comment (Luffbee) [reply server/schedulers/hot_region.go]: What about `hotRegions`? Or `hotPeers`?

Comment (rleungx) [reply server/schedulers/hot_region.go]: Both are better than `storeHots` and I prefer to unify the naming of `HotPeerStat` and `storeHotRegionsDefaultLen`

Comment (Luffbee) [reply server/schedulers/hot_region.go]: Unified similar names.

File: server/statistics/store.go [hunk 5]
```diff
@@ -189,13 +230,42 @@ func (r *RollingStoreStats) Observe(stats *pdpb.StoreStats) {
...
+}
+
 // GetBytesRate returns the bytes write rate and the bytes read rate.
 func (r *RollingStoreStats) GetBytesRate() (writeRate float64, readRate float64) {
 	r.RLock()
 	defer r.RUnlock()
 	return r.bytesWriteRate.Get(), r.bytesReadRate.Get()
 }
 
+// GetBytesWriteRate returns the bytes write rate and the bytes read rate.
```
Comment (rleungx) [inline server/statistics/store.go]: The comment needs to change.

File: server/statistics/store.go [hunk 6]
```diff
@@ -107,6 +128,26 @@ func (s *StoresStats) GetStoreBytesRate(storeID uint64) (writeRate float64, read
...
+func (s *StoresStats) GetStoreBytesWriteRate(storeID uint64) float64 {
+	s.RLock()
+	defer s.RUnlock()
+	if storeStat, ok := s.rollingStoresStats[storeID]; ok {
+		return storeStat.GetBytesWriteRate()
+	}
+	return 0
+}
+
+// GetStoreBytesReadRate returns the bytes write stat of the specified store.
```
Comment (rleungx) [inline server/statistics/store.go]:
```suggestion
// GetStoreBytesReadRate returns the bytes read stat of the specified store.
```

File: server/statistics/store.go [hunk 7]
```diff
@@ -189,13 +230,42 @@ func (r *RollingStoreStats) Observe(stats *pdpb.StoreStats) {
...
 }
 
+// GetBytesWriteRate returns the bytes write rate and the bytes read rate.
+func (r *RollingStoreStats) GetBytesWriteRate() float64 {
+	r.RLock()
+	defer r.RUnlock()
+	return r.bytesWriteRate.Get()
+}
+
+// GetBytesReadRate returns the bytes write rate and the bytes read rate.
```
Comment (rleungx) [inline server/statistics/store.go]: ditto

File: server/statistics/store_hot_peers_infos.go [hunk 8]
```diff
@@ -13,11 +13,11 @@
 
 package statistics
 
-// StoreHotRegionInfos : used to get human readable description for hot regions.
-type StoreHotRegionInfos struct {
-	AsPeer   StoreHotRegionsStat `json:"as_peer"`
-	AsLeader StoreHotRegionsStat `json:"as_leader"`
+// StoreHotPeersInfos : used to get human readable description for hot regions.
```
Comment (rleungx) [inline server/statistics/store_hot_peers_infos.go]:
```suggestion
// StoreHotPeersInfos is used to get human-readable description for hot regions.
```

File: server/statistics/store.go (hunk 5 above)
Comment (Luffbee) [reply server/statistics/store.go]: Fixed.

File: server/statistics/store_hot_peers_infos.go [hunk 9]
```diff
@@ -13,11 +13,11 @@
...
-	AsLeader StoreHotRegionsStat `json:"as_leader"`
+// StoreHotPeersInfos is used to get human-readable description for hot regions.
+type StoreHotPeersInfos struct {
+	AsPeer   StoreHotPeersStat `json:"as_peer"`
+	AsLeader StoreHotPeersStat `json:"as_leader"`
 }
 
-// StoreHotRegionsStat used to record the hot region statistics group by store
-type StoreHotRegionsStat map[uint64]*HotRegionsStat
+// StoreHotPeersStat used to record the hot region statistics group by store
```
Comment (lhy1024) [inline server/statistics/store_hot_peers_infos.go]:
```suggestion
// StoreHotPeersStat is used to record the hot region statistics group by store.
```

File: server/schedulers/hot_test.go [hunk 10]
```diff
@@ -52,85 +53,135 @@ func (s *testHotWriteRegionSchedulerSuite) TestSchedule(c *C) {
...
-	tc.UpdateStorageWrittenBytes(4, 60*1024*1024)
+	//| store_id | write_bytes_rate |
+	//|----------|------------------|
+	//|    1     |       7.5MB      |
+	//|    2     |       4.5MB      |
+	//|    3     |       4.5MB      |
+	//|    4     |        6MB       |
+	//|    5     |        0MB       |
+	//|    6     |        0MB       |
+	tc.UpdateStorageWrittenBytes(1, 7.5*MB*statistics.StoreHeartBeatReportInterval)
```
Comment (lhy1024) [inline server/schedulers/hot_test.go]: `statistics.StoreHeartBeatReportInterval` is too long.

Comment (Luffbee) [reply server/schedulers/hot_test.go]: I noticed that, but I didn't find any better way to solve it.

**Pull request merged**